docs/syntax-if-statements-initializers.md

```diff
@@ -0,0 +1,114 @@
+# Initializers in if statements
+
+# Summary
+
+Introduce an initializer expression that declares and initializes variables in if statements.
+
+# Motivation
+
+Initializers can improve code clarity and reduce scope pollution by allowing developers to declare variables in the conditions of if statements. ⁤⁤ Declaring variables at the point of use in if statements for the scope of the if statement's block simplifies code, leading to better readability and understanding of program logic. By limiting the scope of variables to the condition's pertaining block, the risk of unintended variable reuse and naming conflicts is reduced. ⁤
+
+The reduced scope pollution improves register space in extreme cases (or auto-generated code) where developers have many variables defined and have to work around register limits by reducing the register size. In some scenarios, especially on Roblox, initializing variables within if statements can lead to improved performance and reduced complexity by avoiding unnecessary calls by developers. ⁤ A common paradigm used by Roblox developers is to use `Instance:FindFirstChild` in their condition and then access it afterwards instead of keeping an existing variable around in the new scope, polluting the existing scope.
+
+# Design
+
+If statements with initializers must match (following the Luau grammar) `'if' 'local' bindinglist ['=' explist] 'then'` and `'local' bindinglist ['=' explist] where exp 'then'` syntax. The variables declared by the initializer are only available to the block of that condition and will be undefined to the `elseif` conditions and blocks as well as the `else` block; any code after the if statement won't have the variables defined either.
+
+In the former case, the value of the first declared variable will be checked.
+
+Example:
+
+```lua
+local function foo()
+    return true
+end
+
+if local b = foo() then
+   print(b)
+else
+   print("false")
+end
+```
+
+`Output: true`
+
+In the latter case, the `exp` condition is checked rather than the initializer.
+
+Example:
+
+```lua
+local function foo()
+    return true
+end
+
+if local b = foo() where b == false then
+   print(b)
+else
+   print("true")
+end
+```
+
+`Output: true`
+
+When declaring multiple values inside of a condition, all of the variables will be evaluated.
```

Comment:
Just for clarification, how would the variables be combined when being tested? I misread the example at first, so I assumed all variables are combined with `or`, but it actually implies using `and`.

```lua
if local a, b = true, false then
  print("pass")
else
  print("fail")
end

-- same as this?
if local a, b = true, false where a and b then
  print("pass")
else
  print("fail")
end
```

I want to argue that this evaluation is ambiguous, since there is no explicit operation between the variables.

Reply:
I've come to agree with @vegorov-rbx on making `where` clause required if more than two locals are involved in the initializer. Good example is found here: https://github.com/luau-lang/rfcs/pull/23/files#r1492770178.

Reply:
I think where clauses should just always be a requirement 😰

Reply:
I commented this elsewhere, but I realized this is pretty much a good point to say it in. The desirable behavior IMO is that any variable initialized in an `if local` statement will be tested to be non-`nil`. The direct argument would be that putting the initialization in a _condition_ is an indication that you want it to be bound to some sort of non-`nil` value in the block following it. The forward-looking advantage of this is that it's consistent with how destructuring should behave, e.g. if we allow destructuring tables in local definitions (#24), you could write something like:

```lua
if local {.x, .y, .z} = point then
    -- x, y, z are defined and non-`nil`
end
```

The main downside is that it means the case where you specifically want to test a result for truthiness, rather than non-`nil`ness will require you to write the condition clause, i.e.

```lua
if local cond = bool() then
    -- cond is not `nil`, but could be `true` or `false`
end

if local cond = bool() in cond then
    -- cond is not `nil`, and has been tested `true` by the condition clause
end

if local cond = bool() in not cond then
    -- cond is not `nil`, and has been tested `false` by the condition clause
end
```

I think this is a reasonable decision to make. Requiring `in` is a more conservative choice that is forwards-compatible with this evolution though.